### ISSUE TITLE
macOS 27 beta: MPSGraph scratch-heap overflow on Gemma 4 12B full-attention (head_dim 512) layers + macOS AOT .aimodelc load regression (AIModelError 3)

### ISSUE THREAD

**Issue by john-rocky** [apple/coreai-models#27]
**Two related macOS 27.0 beta Core AI runtime bugs that block 12B-class Gemma 4 on Mac GPU.** A clean one-layer bisection isolates the trigger. Numerics are verified-correct and the graph runs at smaller sizes — both blockers are runtime / MPSGraph-side, not model-side. (Related to #5, another macOS-27-beta MPSGraph lowering bug.)

### Environment
- macOS **27.0** build **26A5353q** (beta), Apple **M4 Max** (`applegpu_g16s`)
- `coreai-build` **3600.67.5.8.1** (MetalToolchain-v27.1.5194)
- coreai-models pipelined engine (`llm-runner` / `llm-benchmark`), `COREAI_CHUNK_THRESHOLD=1`
- Model: a **Gemma 4 12B** dense decode-only pipelined bundle (in-graph embed+head, one growing KV pair, dual head_dim 256 sliding / 512 full, `attention_k_eq_v` full layers)

### Bug 1 — MPSGraph scratch-heap overflow on full-attention (head_dim 512) layers
At the **first decode token** the engine aborts:
```
allocateMTLBufferFromMTLHeap: offset 198400 + size 16384 exceeds heap total 212992
.../MPSRuntime/Operations/GPUMemrefOps.mm:687: failed assertion
  'Failed to acquire the source buffer for the ViewOp'
```
**Decisive bisection:**
- `--num-layers 5` (all *sliding*, head_dim 256) → **runs** (~409 tok/s)
- `--num-layers 6` (adds the first *full* layer: head_dim 512, 16 query heads) → **crashes**

The failing buffer is exactly `[1, 16, 1, 512]` fp16 = **16384 B**, the full layer's `q_proj` output. It scales with the number of full layers (16 KB at 1 full layer, 32 KB at 2) and overflows MPSGraph's ~208 KB decode scratch heap (mis-sized by ~2 KB). Sliding-layer Q (`[1,16,1,256]` = 8 KB) fits, and **Gemma 4 E2B/E4B full layers (8 heads × 512 = 8 KB) also fit and run** — only the 12B's 16-head × 512 Q tips the heap over. The crash is **invariant** to every graph-source change tried (KV cache pad↔replicate, uniform narrow, `.contiguous()` on Q and on K/V, vanilla vs HF SDPA): identical heap / offset / size each time.

### Bug 2 — AOT `.aimodelc` fails to load on macOS (regression vs iOS)
Pre-compiling for the **correct M4 Max arch** succeeds:
```
xcrun coreai-build compile <bundle>.aimodel --platform macOS --architecture h16s --expect-frequent-reshapes -o /tmp/aot
```
…but loading the resulting `.aimodelc` fails:
```
CoreAIDelegates.AIModelError error 3      (raw AIModel.load)
invalidCompiledModel                      (llm-runner / LanguageBundle)
```
This is **not** specific to the Bug-1 graph: a model that JIT-runs perfectly (`--num-layers 5`, all sliding) **also fails to AOT-load** with the same `AIModelError 3`. So this macOS build cannot load *any* precompiled `.aimodelc` for a macOS target, while the **same Core AI runtime loads AOT `.aimodelc` fine on iOS** (h18p bundles run on iPhone 17 Pro). With/without `--expect-frequent-reshapes` and with the source `.aimodel` present alongside, same result.

### Impact
Together these block all GPU paths for 12B-class Gemma 4 (and likely any model whose per-layer attention intermediate exceeds the scratch heap) on Mac: JIT crashes (Bug 1), AOT-load is rejected (Bug 2), and the CPU delegate also fails to load. If Bug 2 were fixed, AOT would work around Bug 1 exactly as iOS already does.

**Comment by andrew9123** [comment]
**Sibling crash? `coreai-pipelined` aborts in `MPSNDArray.mm:893` ("Must be 64 bytes") on Qwen3-0.6B — sequential works on the same bundle**

Posting here since this looks like another macOS-27-beta MPSGraph buffer-sizing regression in the same family as this issue and #5, though the **assertion site and trigger differ** (here it's `MPSNDArray.mm:893` / 64-byte alignment during the pipelined GPU path, not the `GPUMemrefOps.mm:687` scratch-heap ViewOp). Happy to split this into its own issue if you'd prefer — flagging it here first in case it's the same root cause.

### Summary

The **`coreai-pipelined`** engine aborts with a `MPSNDArray` "buffer is not large enough. Must be 64 bytes" assertion when generating, on a stock `coreai.llm.export Qwen/Qwen3-0.6B` bundle. The **`coreai-sequential`** engine runs the **same bundle** fine (~170 tok/s decode). The crash is **invariant to quantization** (4-bit *and* fp16 `--compression none` both abort) and **`--synchronous-sampling` does not help** — which isolates it to the pipelined engine's GPU path (on-GPU MPS sampling / NDArray aliasing), not the export or the weights.

### Environment

- macOS **27.0 (26A5368g)** — beta 2; Apple **M5** (`Mac17,4`), 32 GB
- Xcode **27.0 (27A5209h)**; `coreai-build` **3600.73.1**
- `coreai-core` **1.0.0b1**, `coreai-torch` **0.4.0**, `coreai-opt` **0.2.0**, `coremltools` **9.0**
- `coreai-models` @ `5a0f161`
- Model: `Qwen/Qwen3-0.6B` via `uv run coreai.llm.export` (4-bit dynamic, `metadata_version` 0.2, `max_context_length` 8192; also reproduced with the fp16 `--compression none` export)

### Steps to reproduce

```bash
# Export (both crash on pipelined):
uv run coreai.llm.export Qwen/Qwen3-0.6B                      # 4-bit dynamic, 335 MB
uv run coreai.llm.export Qwen/Qwen3-0.6B --compression none   # fp16, ~1.2 GB

BUNDLE=exported/qwen3-0_6b/qwen3_0_6b_4bit_dynamic

# CRASHES — pipelined (also the auto-selected engine for a dynamic bundle, so a plain run hits this too):
swift run -c release llm-runner --model "$BUNDLE" \
  --prompt "Explain on-device AI in one sentence." --max-tokens 24 \
  --sampling-strategy greedy --inference-engine-variant coreai-pipelined
#   ...also crashes with --synchronous-sampling, and with the fp16 bundle.

# WORKS — same bundle, sequential engine (~214/875/4707 tok/s prefill across 59/192/463-tok prompts, ~156–175 tok/s decode):
swift run -c release llm-runner --model "$BUNDLE" \
  --prompt "Explain on-device AI in one sentence." --max-tokens 24 \
  --sampling-strategy greedy --inference-engine-variant coreai-sequential
```

### Crash output (pipelined)

```
⏳ Preparing AI asset from source.../AppleInternal/Library/BuildRoots/.../Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:893: failed assertion `[MPSNDArray, initWithBufferImpl:offset:descriptor:isForNDArrayAlias:isUserBuffer:] Error: buffer is not large enough. Must be 64 bytes
'
 done in 0.094s

Generating...
[process aborts — SIGABRT, exit 134; no tokens emitted]
```

(The assertion text interleaves with the buffered progress logs; the assertion is the only error emitted before the abort.)

### What this isolates

| Variable | Result |
|---|---|
| `--inference-engine-variant coreai-pipelined` | **crash** |
| `--inference-engine-variant coreai-sequential` (same bundle) | **works** (~170 tok/s decode) |
| 4-bit dynamic vs fp16 (`--compression none`) | **both crash** (pipelined) |
| `--synchronous-sampling` | no effect (still crashes) |

So it's specific to the pipelined engine's GPU path; the export, tokenizer, and weights are fine (sequential loads the same bundle and generates correctly).

### Hypothesis (low confidence — for triage)

`64` is the ANE/Core ML last-axis buffer-alignment floor, and `MPSNDArray … buffer is not large enough` is the classic descriptor-vs-`MTLBuffer` size mismatch (same assertion seen for years in the PyTorch MPS backend on non-contiguous/edge-size tensors). My guess is the pipelined engine's **on-GPU sampling** (`MPSGraphSamplers` argmax/topK) or a 1-token-wide decode-slice view constructs an `MPSNDArray` whose last dimension lands **below the 64-byte minimum**, which this beta's MPS no longer tolerates. The sequential engine samples off-GPU and doesn't build that aliased view, which would explain why it's unaffected, and the quant-invariance points at the sampler/decode path rather than the weights. Could be the same underlying 27-beta MPSGraph buffer-sizing regression as this issue / #5, just surfacing at a different op.


### Feedback Assistant

This has been filed as FB23395590 in Feedback Assistant.

**Comment by john-rocky** [comment]
### New manifestation: the same scratch-heap ceiling also blocks **multi-token prefill on iOS**, for **E2B** — not just 12B decode on macOS

Bug 1 above notes that Gemma 4 E2B/E4B *decode* full layers (8 heads × 512 = 8 KB Q) fit and run. That holds for **S=1 decode**. But the scratch heap is also overflowed by **E2B on iOS during multi-token prefill**, because the per-encode intermediates scale with **query width `S`**, not only head_dim × heads.

**Environment**
- iOS **27.0** build **24A5380h** (public beta = dev beta 3), **iPhone 17 Pro** (`h18p`)
- `coreai-build` **3600.75.3** (MetalToolchain **27A5218h**), `coreai-core 1.0.0b2`
- Model: **Gemma 4 E2B** QAT int4lin `tbl` decode bundle (in-graph PLE gather + head), one growing KV pair. Multifunction export: `main` = S=1 decode + `prefill` = static S=`C` chunk (shared weights), so the engine walks a long prompt in fixed `C`-token chunks.

**Result** — `main` (S=1) decode is numerics-verified (24/24 greedy match). The static S=`C` `prefill` function aborts on the first chunk, and **no chunk size fits** — shrinking `C` just moves which buffer overflows the **145920 B** (~142 KB) heap:

| chunk `C` | abort |
|---|---|
| 64 | `allocateMTLBufferFromMTLHeap: offset 512 + size 196608 exceeds heap total 145920` — the hidden ViewOp `2·C·1536·2` (= 64·1536·2) |
| 32 | `offset 98816 + size 98304 exceeds heap total 145920` — two concurrent `32·1536·2` hidden buffers |
| 16 | clears the hidden buffers, then `offset 98816 + size 573440 exceeds heap total 145920` — a ~560 KB attention intermediate (`GPUMemrefOps.mm:700`, same `ViewOp` assertion) |

So the ceiling is hit by **whichever** per-encode intermediate first exceeds the heap (hidden state `∝ C`, or the attention buffer), and reducing `C` cannot escape it — matching this issue's "invariant to model-side graph squeezes."

**Two extensions to Bug 1 worth noting:**
1. **Platform** — the overflow is **not macOS-only**. The iOS runtime hits the identical `allocateMTLBufferFromMTLHeap … exceeds heap total` / `ViewOp` assertion; the iOS heap here is even smaller (**145920** vs the macOS **212992**).
2. **Path & model class** — it is not limited to 12B decode. Any **multi-token prefill** (`S>1`) on **E2B** overflows, because the intermediates scale with query width. Only **S=1** paths stay under the heap.

**Correctness** — the chunked prefill is numerically correct: a pure-PyTorch parity of one S=`C` forward vs `C` sequential S=1 steps matches to ~2e-5 on logits and KV (argmax-identical). So this is purely runtime/MPSGraph-side, consistent with the original report.

**Practical consequence** — on iOS the only prefill path that stays under the scratch heap is the **per-token provider (S=1) mode**, i.e. batched prefill is unavailable and a 1024-token prompt must be processed one token at a time (prefill throughput ≈ decode throughput). Raising the per-encode scratch-heap ceiling (or the sizing fix noted for Bug 1) would unblock batched prefill on iOS, not just 12B decode on macOS.

**Comment by stikves** [comment]
Thank you for the bug report.

I was able to replicate the overflow issue on an M2 Mac. For others I'll check whether they are related or require independent solutions.

**Comment by stikves** [comment]
Closing as could not replicate on recent macOS beta.

Feel free to re-open if the issue reappears.

**Comment by john-rocky** [comment]
Thanks for the triage pass on this — closing the macOS half makes sense. The iOS manifestation from my July comment still reproduces on the current beta (24A5418b, 2/2 runs, identical offsets/sizes/heap), so I've filed that face as its own issue with today's environment and logs: #201. S=1 paths still run fine.
